cfgrammar/src/lib/analysis.rs

```diff
@@ -0,0 +1,73 @@
+use crate::yacc::ast::{GrammarAST, Symbol};
+use crate::yacc::{YaccGrammarWarning, YaccGrammarWarningKind};
+use lazy_static::lazy_static;
+use std::ops::Deref;
+
+/// Performs an analysis on a given `Subject`, the given
+/// subject may be known to have pre-existing errors.
+pub trait Analysis<Subject> {
```

Comment:
I'm not sure I understand why we need such a generic struct. Any suggestions?

[I'd also prefer `fn analyse` as the spelling is more congruent with `trait Analysis`. But this is perhaps just my inner British spelling nerd coming out...]

Reply:
The main reason it's generic is so it can be used when the `Subject`s are declared in the various crates.
I couldn't really think of anything else which fit, given the limitations.

1. It's owned/taking self by mutable ref so instead of returning it could be used in the builder call chains.
2. It's generic on the ast/grammar/table because those all live in different crates.
3. Then does the work via side-effect to extract an analysis specific type to avoid imposing a return type on every analysis.

I really couldn't think of anything else that would cover those 3 constraints, the last seemed like the most important, 
The other 2 seemed somewhat necessary to achieving that.

Anyhow that was overall the reasoning behind the trait, I tend to think it is pretty strange trait though.

I hadn't thought of the spelling difference, but am happy to change it.

Reply:
So, I think one thing which might work instead of this generic analysis trait, is make `analyse_*` take closures,

```
fn analyse_ast<T>(t: T, f: &dyn Fn(GrammarAST, T)) {
    f(GrammarAST, t)
}

  let mut warnings: Vec<YaccGrammarWarning> = Vec::new();
  analyse_ast(&mut warnings, &|ast, ws| { 
      ws.push(YaccGrammarWarning)
  })
```

At least I believe that could be one way to get rid of the trait.

Reply:
I think part of my confusion is that I think I only see one `impl` of this `trait`?

Reply:
There is a total of 4, throughout the series,
2. [`CTConflictsAnalysis`](https://github.com/softdevteam/grmtools/pull/353/commits/fd683fe9b3e27c4aea87979ce59aac830c8dacb0#diff-f389393ec514f9efce5aaf9e5188b4c3f2628445cd5fefad72e0006fb7334930R1315-R1331)

2 in the example:

3. [`AriadneWarningAnalysis`](https://github.com/softdevteam/grmtools/pull/353/commits/440754e50622276e3f63183375c503464c6b0057#diff-cb061aa1540e6f5a20a8464a3739c334243bf0957ea32b18419e0cc7eafe40b9R100-R104) -- which just wraps the YaccGrammarAnalysis
4. [`EmptyAnalysis`](https://github.com/softdevteam/grmtools/pull/353/commits/440754e50622276e3f63183375c503464c6b0057#diff-cb061aa1540e6f5a20a8464a3739c334243bf0957ea32b18419e0cc7eafe40b9R17-R19) -- used to just show that multiple analysis can be run and the places it can be used

There should be one in my project too which has a mpsc channel field.

Reply:
Ah, sorry, I didn't search for the right string! Let me take a gander.

Reply:
I recall now, that I have one more thing for generating railroad diagrams, which I intend to convert to an analysis,
I could quite easily add it to the example, It would likely mean we should rename the pretty_errors example, and it adds some dependencies though.

Anyhow, let me know if you'd like another.